assets/js/hocs/test/with-product.js

```diff
@@ -0,0 +1,95 @@
+/**
+ * External dependencies
+ */
+import TestRenderer from 'react-test-renderer';
+
+/**
+ * Internal dependencies
+ */
+import withProduct from '../with-product';
+import * as mockUtils from '../../components/utils';
+
+// Mock the getProduct functions for tests.
+jest.mock( '../../components/utils', () => ( {
+	getProduct: jest.fn().mockImplementation(
+		() => Promise.resolve()
+	),
+} ) );
+
+const mockProduct = { name: 'T-Shirt' };
+const attributes = { productId: 1 };
+const TestComponent = withProduct( ( props ) => {
+	return <div
+		error={ props.error }
+		getProduct={ props.getProduct }
+		isLoading={ props.isLoading }
+		product={ props.product }
+	/>;
+} );
+const render = () => {
+	return TestRenderer.create(
+		<TestComponent
+			attributes={ attributes }
+		/>
+	);
+};
+
+describe( 'withProduct Component', () => {
+	afterEach( () => {
+		mockUtils.getProduct.mockClear();
+	} );
+
+	describe( 'lifecycle events', () => {
+		const renderer = render();
+
+		it( 'getProduct is called on mount with passed in product id', () => {
+			const { getProduct } = mockUtils;
+
+			expect( getProduct ).toHaveBeenCalledWith( attributes.productId );
+			expect( getProduct ).toHaveBeenCalledTimes( 1 );
+		} );
+
+		it( 'getProduct is hooked to the prop', () => {
+			const { getProduct } = mockUtils;
+			const props = renderer.root.findByType( 'div' ).props;
+
+			props.getProduct();
+
+			expect( getProduct ).toHaveBeenCalledTimes( 1 );
+		} );
+	} );
+
+	describe( 'when the API returns product data', () => {
+		mockUtils.getProduct.mockImplementation(
+			( productId ) => Promise.resolve( { ...mockProduct, id: productId } )
+		);
+		const renderer = render();
+
+		it( 'sets the product props', () => {
+			const props = renderer.root.findByType( 'div' ).props;
+
+			expect( props.error ).toEqual( null );
```

Comment:
Nitpick: for assertions like this (primitives, null) it's good to use the more strict `toBe` assertion (tad bit faster) because it'll assert type as well.

Reply:
Good catch! Fixed in a6f52e51842717f7a061f080dd807acd72848d23.